dev/src/dev/debug_qp.clj

```diff
@@ -14,6 +14,85 @@
             [metabase.util :as u]
             [toucan.db :as db]))
 
+;;;; [[->sorted-mbql-query-map]]
+
+(def ^:private mbql-clause->sort-order
+  (into {}
+        (map-indexed (fn [i k]
+                       [k i]))
+        [;; top-level keys
+         :database
+         :type
+         :query
+         :native
+         ;; inner-query and join keys
+         :source-table
+         :source-query
+         :source-metadata
+         :alias
+         :joins
+         :expressions
+         :breakout
+         :aggregation
+         :condition
+         :fields
+         :strategy
+         :filter
+         :order-by
+         :page
+         :limit]))
+
+(defn- sorted-mbql-query-map []
+  ;; stuff in [[mbql-clause->sort-order]] should always get sorted according to that order. Everything else should go at
+  ;; the end, with non-namespaced stuff first and namespaced stuff last; otherwise sort alphabetically
+  (sorted-map-by (fn [x y]
+                   (let [order   (fn [k]
+                                   (or (mbql-clause->sort-order k)
+                                       (when (and (keyword? k) (namespace k))
+                                         Integer/MAX_VALUE)
+                                       (dec Integer/MAX_VALUE)))
+                         x-order (order x)
+                         y-order (order y)]
+                     (if (= x-order y-order)
+                       (compare (str x) (str y))
+                       (compare x-order y-order))))))
+
+(def ^:dynamic *shorten-namespaced-keywords?*
```

Comment:
stuff like this makes me think we should just not use auto-resolved namespace keywords at all. `:add/source-info` would be the same information, not likely to clash with anything, and would not risk errors in the reader.

Reply:
True, but there are a few benefits to them: 
- they're basically guaranteed to be unique 
- it's super clear where they came from
- to get auto-resolution you have to `require` the namespace that introduces them which is usually desirable from a topological namespace dependency standpoint e.g. it enforces the constraint that `add-alias-info` does not depend on any of the namespaces that use the `::add/` keys 

RE reader errors: This only shortens stuff if there's an alias inside `*ns*`. So that shouldn't cause any reader errors, unless you take that form and try to evaluate it in a different namespace that doesn't have that alias. There's actually no good reason I can think of to turn this switch off, but I put it there just in case something comes up in the future

Reply:
FWIW I think you're right about just giving them some sort of arbitrary prefix so we don't have to mess around with weird things -- in hindsight, `:add/` or something like that would have probably been a better namespace than `::add/` for the reasons you mentioned. Not sure how easy this is going to be to change tho